{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting Flask-User<0.7\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/cc/00/7745a03b29fc45395efefaf5ce59162c2de13a0c895fafb0d57ddada2503/Flask-User-0.6.21.tar.gz (64kB)\n",
      "\u001b[K     |████████████████████████████████| 71kB 681kB/s eta 0:00:01\n",
      "\u001b[?25hCollecting bcrypt (from Flask-User<0.7)\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/62/20/4c94f3f8dfc6b8720c8bc903ce2951ec6397ad864e3a64b4abdced014514/bcrypt-3.1.7-cp34-abi3-macosx_10_6_intel.whl (53kB)\n",
      "\u001b[K     |████████████████████████████████| 61kB 1.4MB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: Flask in /Applications/anaconda3/lib/python3.7/site-packages (from Flask-User<0.7) (1.1.1)\n",
      "Requirement already satisfied: Flask-Login in /Applications/anaconda3/lib/python3.7/site-packages (from Flask-User<0.7) (0.5.0)\n",
      "Collecting Flask-Mail (from Flask-User<0.7)\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/05/2f/6a545452040c2556559779db87148d2a85e78a26f90326647b51dc5e81e9/Flask-Mail-0.9.1.tar.gz (45kB)\n",
      "\u001b[K     |████████████████████████████████| 51kB 2.6MB/s eta 0:00:011\n",
      "\u001b[?25hCollecting Flask-SQLAlchemy (from Flask-User<0.7)\n",
      "  Downloading https://files.pythonhosted.org/packages/1e/65/226d95466c75e34e291a76890ed0e27af2e46ab913002847856f11d4d59d/Flask_SQLAlchemy-2.4.1-py2.py3-none-any.whl\n",
      "Collecting Flask-WTF (from Flask-User<0.7)\n",
      "  Downloading https://files.pythonhosted.org/packages/36/a9/8c01171066bd7a524ee005d81bb4a8aa446ab178043a1ad6cb5dc8f0bd83/Flask_WTF-0.14.3-py2.py3-none-any.whl\n",
      "Collecting passlib (from Flask-User<0.7)\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/11/b8/e9a78f3033228013ba8564adad8d0031bf9d39ea3acc3cdb9d55fabeb4ba/passlib-1.7.2-py2.py3-none-any.whl (507kB)\n",
      "\u001b[K     |████████████████████████████████| 512kB 1.4MB/s eta 0:00:01\n",
      "\u001b[?25hCollecting pycryptodome (from Flask-User<0.7)\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/0e/b9/6cc7e451a7b35d9151d52ee904525aa8b883489095bfc61c3d45003c4f47/pycryptodome-3.9.7-cp37-cp37m-macosx_10_6_intel.whl (14.1MB)\n",
      "\u001b[K     |████████████████████████████████| 14.1MB 169kB/s eta 0:00:01    |▊                               | 296kB 326kB/s eta 0:00:43     |██████▋                         | 2.9MB 187kB/s eta 0:01:00\n",
      "\u001b[?25hRequirement already satisfied: six>=1.4.1 in /Applications/anaconda3/lib/python3.7/site-packages (from bcrypt->Flask-User<0.7) (1.12.0)\n",
      "Requirement already satisfied: cffi>=1.1 in /Applications/anaconda3/lib/python3.7/site-packages (from bcrypt->Flask-User<0.7) (1.12.3)\n",
      "Requirement already satisfied: itsdangerous>=0.24 in /Applications/anaconda3/lib/python3.7/site-packages (from Flask->Flask-User<0.7) (1.1.0)\n",
      "Requirement already satisfied: click>=5.1 in /Applications/anaconda3/lib/python3.7/site-packages (from Flask->Flask-User<0.7) (7.0)\n",
      "Requirement already satisfied: Werkzeug>=0.15 in /Applications/anaconda3/lib/python3.7/site-packages (from Flask->Flask-User<0.7) (0.16.0)\n",
      "Requirement already satisfied: Jinja2>=2.10.1 in /Applications/anaconda3/lib/python3.7/site-packages (from Flask->Flask-User<0.7) (2.10.3)\n",
      "Collecting blinker (from Flask-Mail->Flask-User<0.7)\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/1b/51/e2a9f3b757eb802f61dc1f2b09c8c99f6eb01cf06416c0671253536517b6/blinker-1.4.tar.gz (111kB)\n",
      "\u001b[K     |████████████████████████████████| 112kB 332kB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: SQLAlchemy>=0.8.0 in /Applications/anaconda3/lib/python3.7/site-packages (from Flask-SQLAlchemy->Flask-User<0.7) (1.3.9)\n",
      "Collecting WTForms (from Flask-WTF->Flask-User<0.7)\n",
      "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/9f/c8/dac5dce9908df1d9d48ec0e26e2a250839fa36ea2c602cc4f85ccfeb5c65/WTForms-2.2.1-py2.py3-none-any.whl (166kB)\n",
      "\u001b[K     |████████████████████████████████| 174kB 251kB/s eta 0:00:01\n",
      "\u001b[?25hRequirement already satisfied: pycparser in /Applications/anaconda3/lib/python3.7/site-packages (from cffi>=1.1->bcrypt->Flask-User<0.7) (2.19)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /Applications/anaconda3/lib/python3.7/site-packages (from Jinja2>=2.10.1->Flask->Flask-User<0.7) (1.1.1)\n",
      "Building wheels for collected packages: Flask-User, Flask-Mail, blinker\n",
      "  Building wheel for Flask-User (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for Flask-User: filename=Flask_User-0.6.21-py2.py3-none-any.whl size=129079 sha256=29b046166d4dc9a9ebe98289d73f5e1775894992b16761ffda07f2d818d5dfcb\n",
      "  Stored in directory: /Users/AzureXu/Library/Caches/pip/wheels/2b/50/f8/77a1837a22279b820e713d4832352d13de8a0576556aac02a9\n",
      "  Building wheel for Flask-Mail (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for Flask-Mail: filename=Flask_Mail-0.9.1-cp37-none-any.whl size=7568 sha256=9dc13b931f731ae1ab8123e38098cfb17972aaeca335eae683cdd67e5aefc722\n",
      "  Stored in directory: /Users/AzureXu/Library/Caches/pip/wheels/eb/aa/d9/34b8f2f9bce7d06a4d07fd46078770584d5504949ebfa286f5\n",
      "  Building wheel for blinker (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for blinker: filename=blinker-1.4-cp37-none-any.whl size=13450 sha256=e8621de0af032342629fc3afaeece6327199d921f05245676478eb6cd735460e\n",
      "  Stored in directory: /Users/AzureXu/Library/Caches/pip/wheels/92/a0/00/8690a57883956a301d91cf4ec999cc0b258b01e3f548f86e89\n",
      "Successfully built Flask-User Flask-Mail blinker\n",
      "Installing collected packages: bcrypt, blinker, Flask-Mail, Flask-SQLAlchemy, WTForms, Flask-WTF, passlib, pycryptodome, Flask-User\n",
      "Successfully installed Flask-Mail-0.9.1 Flask-SQLAlchemy-2.4.1 Flask-User-0.6.21 Flask-WTF-0.14.3 WTForms-2.2.1 bcrypt-3.1.7 blinker-1.4 passlib-1.7.2 pycryptodome-3.9.7\n"
     ]
    }
   ],
   "source": [
    "!pip install \"Flask-User<0.7\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " * Running on http://localhost:9030/ (Press CTRL+C to quit)\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:16] \"GET / HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:16] \"GET /static/css/bootstrap.min.css HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:16] \"GET /static/js/jquery.min.js HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:16] \"GET /static/js/bootstrap.min.js HTTP/1.1\" 200 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:17] \"GET /favicon.ico HTTP/1.1\" 404 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:17] \"GET /static/fonts/glyphicons-halflings-regular.woff2 HTTP/1.1\" 200 -\n",
      "[2020-03-27 02:12:23,091] ERROR in app: Exception on /album/1 [GET]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/_collections.py\", line 1010, in __call__\n",
      "    return self.registry[key]\n",
      "KeyError: <greenlet.greenlet object at 0x10c1a53c0>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_cors/extension.py\", line 161, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-18-d28221e7accf>\", line 92, in album_paging\n",
      "    nopic = locations_in_HK.query.paginate(per_page=3, page=page_num, error_out=True)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 519, in __get__\n",
      "    return type.query_class(mapper, session=self.sa.session())\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/orm/scoping.py\", line 78, in __call__\n",
      "    return self.registry()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/_collections.py\", line 1012, in __call__\n",
      "    return self.registry.setdefault(key, self.createfunc())\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/orm/session.py\", line 3225, in __call__\n",
      "    return self.class_(**local_kw)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 138, in __init__\n",
      "    bind = options.pop('bind', None) or db.engine\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 937, in engine\n",
      "    return self.get_engine()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 956, in get_engine\n",
      "    return connector.get_engine()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 561, in get_engine\n",
      "    self._engine = rv = self._sa.create_engine(sa_url, options)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 966, in create_engine\n",
      "    return sqlalchemy.create_engine(sa_url, **engine_opts)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/__init__.py\", line 479, in create_engine\n",
      "    return strategy.create(*args, **kwargs)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/strategies.py\", line 87, in create\n",
      "    dbapi = dialect_cls.dbapi(**dbapi_args)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/dialects/mysql/mysqldb.py\", line 118, in dbapi\n",
      "    return __import__(\"MySQLdb\")\n",
      "ModuleNotFoundError: No module named 'MySQLdb'\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:23] \"GET /album/1 HTTP/1.1\" 500 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:23] \"GET /favicon.ico HTTP/1.1\" 404 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:24] \"GET /list_album HTTP/1.1\" 404 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:25] \"GET /favicon.ico HTTP/1.1\" 404 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:33] \"GET / HTTP/1.1\" 200 -\n",
      "[2020-03-27 02:12:33,820] ERROR in app: Exception on /album/1 [GET]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/_collections.py\", line 1010, in __call__\n",
      "    return self.registry[key]\n",
      "KeyError: <greenlet.greenlet object at 0x10c1a53c0>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_cors/extension.py\", line 161, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-18-d28221e7accf>\", line 92, in album_paging\n",
      "    nopic = locations_in_HK.query.paginate(per_page=3, page=page_num, error_out=True)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 519, in __get__\n",
      "    return type.query_class(mapper, session=self.sa.session())\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/orm/scoping.py\", line 78, in __call__\n",
      "    return self.registry()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/_collections.py\", line 1012, in __call__\n",
      "    return self.registry.setdefault(key, self.createfunc())\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/orm/session.py\", line 3225, in __call__\n",
      "    return self.class_(**local_kw)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 138, in __init__\n",
      "    bind = options.pop('bind', None) or db.engine\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 937, in engine\n",
      "    return self.get_engine()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 956, in get_engine\n",
      "    return connector.get_engine()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 561, in get_engine\n",
      "    self._engine = rv = self._sa.create_engine(sa_url, options)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 966, in create_engine\n",
      "    return sqlalchemy.create_engine(sa_url, **engine_opts)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/__init__.py\", line 479, in create_engine\n",
      "    return strategy.create(*args, **kwargs)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/strategies.py\", line 87, in create\n",
      "    dbapi = dialect_cls.dbapi(**dbapi_args)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/dialects/mysql/mysqldb.py\", line 118, in dbapi\n",
      "    return __import__(\"MySQLdb\")\n",
      "ModuleNotFoundError: No module named 'MySQLdb'\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:33] \"GET /album/1 HTTP/1.1\" 500 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:34] \"GET / HTTP/1.1\" 200 -\n",
      "[2020-03-27 02:12:39,931] ERROR in app: Exception on /album/1 [GET]\n",
      "Traceback (most recent call last):\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/_collections.py\", line 1010, in __call__\n",
      "    return self.registry[key]\n",
      "KeyError: <greenlet.greenlet object at 0x10c1a53c0>\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 2446, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1951, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_cors/extension.py\", line 161, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1820, in handle_user_exception\n",
      "    reraise(exc_type, exc_value, tb)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/_compat.py\", line 39, in reraise\n",
      "    raise value\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1949, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask/app.py\", line 1935, in dispatch_request\n",
      "    return self.view_functions[rule.endpoint](**req.view_args)\n",
      "  File \"<ipython-input-18-d28221e7accf>\", line 92, in album_paging\n",
      "    nopic = locations_in_HK.query.paginate(per_page=3, page=page_num, error_out=True)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 519, in __get__\n",
      "    return type.query_class(mapper, session=self.sa.session())\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/orm/scoping.py\", line 78, in __call__\n",
      "    return self.registry()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/_collections.py\", line 1012, in __call__\n",
      "    return self.registry.setdefault(key, self.createfunc())\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/orm/session.py\", line 3225, in __call__\n",
      "    return self.class_(**local_kw)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 138, in __init__\n",
      "    bind = options.pop('bind', None) or db.engine\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 937, in engine\n",
      "    return self.get_engine()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 956, in get_engine\n",
      "    return connector.get_engine()\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 561, in get_engine\n",
      "    self._engine = rv = self._sa.create_engine(sa_url, options)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py\", line 966, in create_engine\n",
      "    return sqlalchemy.create_engine(sa_url, **engine_opts)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/__init__.py\", line 479, in create_engine\n",
      "    return strategy.create(*args, **kwargs)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/strategies.py\", line 87, in create\n",
      "    dbapi = dialect_cls.dbapi(**dbapi_args)\n",
      "  File \"/Applications/anaconda3/lib/python3.7/site-packages/sqlalchemy/dialects/mysql/mysqldb.py\", line 118, in dbapi\n",
      "    return __import__(\"MySQLdb\")\n",
      "ModuleNotFoundError: No module named 'MySQLdb'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "127.0.0.1 - - [27/Mar/2020 02:12:39] \"GET /album/1 HTTP/1.1\" 500 -\n",
      "127.0.0.1 - - [27/Mar/2020 02:12:41] \"GET /list_album HTTP/1.1\" 404 -\n"
     ]
    }
   ],
   "source": [
    "############# Flask Modules Setup ##############\n",
    "\n",
    "from flask import Flask, render_template, url_for, redirect, request, make_response, Response, jsonify\n",
    "from flask_login import LoginManager, login_user, current_user, logout_user, login_required, UserMixin\n",
    "from flask_cors import CORS\n",
    "from flask_sqlalchemy import SQLAlchemy\n",
    "from datetime import datetime\n",
    "import requests\n",
    "\n",
    "############ Initialize Flask App ##############\n",
    "\n",
    "app = Flask(__name__)\n",
    "CORS(app)\n",
    "\n",
    "#### MySQL SQLAlchemy Object Relations Mapping #####\n",
    "\n",
    "app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql://root:root@localhost/locations_in_HK'\n",
    "app.config['SECRET_KEY'] = \"mysecret\"\n",
    "app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False\n",
    "db = SQLAlchemy(app)\n",
    "\n",
    "class locations_in_HK(db.Model):\n",
    "    __tablename__ = 'nopic'\n",
    "    id = db.Column(db.Integer,primary_key=True)\n",
    "    Films = db.Column(db.String)\n",
    "    poster= db.Column(db.String(30))\n",
    "    locations = db.Column(db.String)\n",
    "    scenetime = db.Column(db.String)\n",
    "    characters = db.Column(db.String)\n",
    "    sceneplot = db.Column(db.String)\n",
    "    # description = db.Column(db.Text)\n",
    "\n",
    "############## Login Manager Setup ###############\n",
    "\n",
    "login_manager = LoginManager()\n",
    "login_manager.init_app(app)\n",
    "login_manager.login_view = 'login'\n",
    "app.config['SECRET_KEY'] = \"lkkajdghdadkglajkgah\"\n",
    "\n",
    "\n",
    "@login_manager.user_loader\n",
    "def load_user(user_id):\n",
    "    return User(user_id)\n",
    "\n",
    "class User(UserMixin):\n",
    "    def __init__(self,id):\n",
    "        self.id = id\n",
    "\n",
    "############ Web Page Routes Setup ###############\n",
    "\n",
    "@app.route(\"/\")\n",
    "def home():\n",
    "    return render_template('home.html')\n",
    "\n",
    "@app.route(\"/login\")\n",
    "def login():\n",
    "    if current_user.is_authenticated:\n",
    "        return redirect(url_for(\"list_album\"))\n",
    "    message = 'Please login in first.'\n",
    "    return render_template('login.html', message=message)\n",
    "\n",
    "@app.route(\"/form_auth\",methods=['POST'])\n",
    "def form_auth():\n",
    "    username = request.form['email']\n",
    "    password = request.form['pwd']\n",
    "    if username == \"lalala@gmail.com\" and password == \"oops\":\n",
    "        login_user(User(1))\n",
    "        # message = \"Dear \" + username + \", your login has been granted. Enjoy your visit !\"\n",
    "        return redirect(url_for(\"list_album\"))\n",
    "    else:\n",
    "        message = 'Wrong password! Please check again'\n",
    "        return render_template('login.html',message=message)\n",
    "\n",
    "@app.route(\"/list_album2\")\n",
    "@login_required\n",
    "def list_album():\n",
    "    dataset = []\n",
    "    nopic_list = locations_in_HK.query.all()\n",
    "    for nopic in nopic_list:\n",
    "        dataset.append({'id':locations_in_HK.id,'Films': locations_in_HK.Films, 'poster': locations_in_HK.poster,\n",
    "                       'locations':locations_in_HK.locations,'scenetime': locations_in_HK.scenetime,'characters': locations_in_HK.characters,'sceneplot': locations_in_HK.sceneplot})\n",
    "    return render_template('list_album2.html', entries=dataset)\n",
    "\n",
    "@app.route('/album')\n",
    "def album():\n",
    "    page_num = 1\n",
    "    nopic = locations_in_HK.query.paginate(per_page=9, page=page_num, error_out=True)\n",
    "    return render_template('product_paging.html', nopic=nopic)\n",
    "\n",
    "@app.route('/album/<int:page_num>')\n",
    "def album_paging(page_num):\n",
    "    nopic = locations_in_HK.query.paginate(per_page=3, page=page_num, error_out=True)\n",
    "    return render_template('product_paging.html', nopic=nopic)\n",
    "\n",
    "@app.route(\"/add_product\",methods=['POST'])\n",
    "@login_required\n",
    "def add_product():\n",
    "    id = request.form['id']\n",
    "    Films = request.form['Films']\n",
    "    poster = request.form['poster']\n",
    "    nopic.locations = request.form['locations']\n",
    "    nopic.scenetime = float(request.form['scenetime'])\n",
    "    nopic.characters = request.form['characters']\n",
    "    nopic.sceneplot = request.form['sceneplot']\n",
    "    nopic = locations_in_HK(Films=Films,poster=poster,locations=locations,scenetime=scenetime,characters=characters,ceneplot=ceneplot)\n",
    "    db.session.add(activity)\n",
    "    db.session.commit()\n",
    "    dataset = []\n",
    "    nopic_list = locations_in_HK.query.all()\n",
    "    for nopic in nopic_list:\n",
    "        dataset.append({'id':locations_in_HK.id,'Films': locations_in_HK.Films, 'poster': locations_in_HK.poster,\n",
    "                       'locations':locations_in_HK.locations,'scenetime': locations_in_HK.scenetime,'characters': locations_in_HK.characters,'sceneplot': locations_in_HK.sceneplot})\n",
    "    return render_template('list_album2.html', entries=dataset)\n",
    "\n",
    "@app.route(\"/update_product\",methods=['POST','PUT'])\n",
    "@login_required\n",
    "def update_product():\n",
    "    record_id = request.form['record_id']\n",
    "    nopic = locations_in_HK.query.filter_by(id=record_id).first()\n",
    "    Films = request.form['Films']\n",
    "    poster = request.form['poster']\n",
    "    nopic.locations = request.form['locations']\n",
    "    nopic.scenetime = float(request.form['scenetime'])\n",
    "    nopic.characters = request.form['characters']\n",
    "    nopic.sceneplot = request.form['sceneplot']\n",
    "    db.session.commit()\n",
    "    dataset = []\n",
    "    nopic_list = locations_in_HK.query.all()\n",
    "    for nopic in nopic_list:\n",
    "        dataset.append({'id':locations_in_HK.id,'Films': locations_in_HK.Films, 'poster': locations_in_HK.poster,\n",
    "                       'locations':locations_in_HK.locations,'scenetime': locations_in_HK.scenetime,'characters': locations_in_HK.characters,'sceneplot': locations_in_HK.sceneplot})\n",
    "    return render_template('list_album2.html', entries=dataset)\n",
    "\n",
    "@app.route(\"/delete_product\",methods=['POST','DELETE'])\n",
    "@login_required\n",
    "def delete_product():\n",
    "    record_id = request.form['record_id']\n",
    "    nopic = locations_in_HK.query.filter_by(id=record_id).first()\n",
    "    db.session.delete(nopic)\n",
    "    db.session.commit()\n",
    "    dataset = []\n",
    "    nopic_list = locations_in_HK.query.all()\n",
    "    for nopic in nopic_list:\n",
    "        dataset.append({'id':locations_in_HK.id,'Films': locations_in_HK.Films, 'poster': locations_in_HK.poster,\n",
    "                       'locations':locations_in_HK.locations,'scenetime': locations_in_HK.scenetime,'characters': locations_in_HK.characters,'sceneplot': locations_in_HK.sceneplot})\n",
    "    return render_template('list_album2.html', entries=dataset)\n",
    "\n",
    "@app.route('/api')\n",
    "def api():\n",
    "    result = db.engine.execute('select * from activities')\n",
    "    final_result = [list(i) for i in result]\n",
    "    dataset=[]\n",
    "    dict={}\n",
    "    for i in final_result:\n",
    "        dict['id'] = i[0]\n",
    "        dict['Films'] = i[1]\n",
    "        dict['poster'] = i[2]\n",
    "        dict['locations'] = i[3]\n",
    "        dict['scenetime'] = i[4]\n",
    "        dict['characters'] = i[5]\n",
    "        dict['sceneplot'] = i[6]\n",
    "        # print(i)\n",
    "        # print(dict)\n",
    "        dataset.append(dict.copy()) #markers.append(fld.copy())\n",
    "    return jsonify({'Album': dataset})\n",
    "\n",
    "@app.route('/logout/')\n",
    "@login_required\n",
    "def logout():\n",
    "    logout_user()\n",
    "    message = 'Thanks for logging out.'\n",
    "    return render_template('login.html',message=message)\n",
    "\n",
    "@app.errorhandler(500)\n",
    "def internal_error(error):\n",
    "    message = 'Wrong password! Please check it again.'\n",
    "    return render_template('login.html',message=message),500\n",
    "\n",
    "######### API Endpoints ##########\n",
    "\n",
    "\n",
    "######### Run Flask Web App at Port 9030 ##########\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    from werkzeug.serving import run_simple\n",
    "    run_simple('localhost', 9030, app)\n",
    "    \n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
